### PR TITLE
Adapt fully_shard

### PR DESCRIPTION
**Pull request by Xing-lil**
### PR Category

User Experience

### PR Types

Others

### Description

- Add support for the `fully_shard` API in PaddleFleet
- Support both `manual_parallel` and `auto_parallel`. 
- The `fully_shard` API no longer accepts the `optimizer`.
- Check on GLM (manual_parallel) and llama3 (auto_parallel)  using `fully_shard` API in PaddleFormers, check same loss, memory, performance compared to the original sharding stage3.

card-92763

### CODE REVIEW

File: test/auto_parallel/semi_auto_parallel_fsdp.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+                stage_losses.append(tr_loss._md5sum())
+        os.environ["skip_sharding3_output_reshard"] = "0"
+        return stage_losses
+
+    def test_fsdp(self):
+        paddle.seed(self._seed)
+        linear = paddle.nn.Linear(10, 10)
+        batch = paddle.rand(shape=[10, 10])
+        # shard the input by sharding degree
+        batch = dist.shard_tensor(batch, self._mesh, [dist.Shard(0)])
```
Comment (liym27) [inline test/auto_parallel/semi_auto_parallel_fsdp.py]: 这里不建议显式调用 shard_tensor

File: test/auto_parallel/semi_auto_parallel_fsdp.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+                tr_loss = linear(batch)
+                tr_loss.backward()
+                tr_loss_add += tr_loss
+                if (step + 1) % self.gradient_accumulation_steps == 0:
+                    tr_loss_add /= self.gradient_accumulation_steps
+                    tr_loss = tr_loss_add
+                    opt.step()
+                    opt.clear_grad()
+                stage_losses.append(tr_loss._md5sum())
+        os.environ["skip_sharding3_output_reshard"] = "0"
```
Comment (liym27) [inline test/auto_parallel/semi_auto_parallel_fsdp.py]: skip_sharding3_output_reshard 是什么作用？需要用户指定吗？

File: test/auto_parallel/semi_auto_parallel_fsdp.py (hunk 1 above)
Comment (Xing-lil) [reply test/auto_parallel/semi_auto_parallel_fsdp.py]: 已改为使用shard_dataloader，感谢！

File: test/auto_parallel/semi_auto_parallel_fsdp.py (hunk 2 above)
Comment (Xing-lil) [reply test/auto_parallel/semi_auto_parallel_fsdp.py]: 是动半sharding3中的优化逻辑，不对用户暴露，已经去除，感谢！

**Review by XieYunshen — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM
